aodncore/pipeline/steps/notify.py

```diff
@@ -277,7 +277,12 @@ def _construct_message(self, recipient_addresses, subject, from_address):
         return message
 
     def _send(self, recipient_addresses, message):
-        smtp_server = smtplib.SMTP(timeout=60)
+        try:
+            smtp_server = smtplib.SMTP(host=self._config.pipeline_config['mail']['smtp_server'],
+                                       port=self._config.pipeline_config['mail'].get('smtp_port', 587),
+                                       timeout=60)
+        except Exception as e:
+            self._logger.warning("Exception thrown when instantiating connection. {e}".format(e=format_exception(e)))
```

Comment:
### Codecov checks

I ran unit tests of this PR on my machine, with a dedicated virtual env just for this repo to install the defined `pip` packages (via `conda`), and have no failed tests, so I'm confused why your machine shows up with some errors?

For the `failed` checks reported by `Codecov` 

![](https://i.imgur.com/XoTY9fi.png)

They're not really failed unit tests,they're just because the 2 new lines to handle exception aren't covered in existing unit tests

https://github.com/aodn/python-aodncore/blob/71262dc85fd24abbc29b74612e30c223f1b616ae/aodncore/pipeline/steps/notify.py#L284-L285

To verify this theory, I made a draft PR using the same changes in this PR but excluded the `exception`, and `Codecov` checks went well.

[Draft PR](https://github.com/aodn/python-aodncore/pull/256)

![](https://i.imgur.com/px8yXaY.png)

----------------

### Code changes

The parameters declared for initialising `smtp_server` will make duplicated declaration with the one under

https://github.com/aodn/python-aodncore/blob/71262dc85fd24abbc29b74612e30c223f1b616ae/aodncore/pipeline/steps/notify.py#L288-L289

while the change in this PR fixed the problem, I think we can refactor the `_send()` method and  initlise `SMTP` client within the `with` statement as described in the 'manufacture' instruction:

https://docs.python.org/3/library/smtplib.html#smtplib.SMTP.connect

> class smtplib.SMTP(host='', port=0, local_hostname=None, [timeout, ]source_address=None) An SMTP instance encapsulates an SMTP connection. It has methods that support a full repertoire of SMTP and ESMTP operations. If the optional host and port parameters are given, the SMTP connect() method is called with those parameters during initialization.

So the proposed refactored code for `_send()` method could be:

```python
def _send(self, recipient_addresses, message):
    with smtplib.SMTP(host=self._config.pipeline_config['mail']['smtp_server'],
                      port=self._config.pipeline_config['mail'].get('smtp_port', 587),
                      timeout=60) as smtp_server:
        sendmail_result = None
        try:
            if self._config.pipeline_config['mail'].get('smtp_tls', True):
                smtp_server.starttls()
            smtp_server.login(self._config.pipeline_config['mail']['smtp_user'],
                              self._config.pipeline_config['mail']['smtp_pass'])
            sendmail_result = smtp_server.sendmail(self._config.pipeline_config['mail']['from'], recipient_addresses,
                                                   message.as_string())
        finally:
            try:
                smtp_server.quit()
            except smtplib.SMTPServerDisconnected:
                pass
            except Exception as e:
                self._logger.warning("exception thrown when closing SMTP. {e}".format(e=format_exception(e)))
        return sendmail_result
```
but this change requires some other updates to the unit tests as well. I need your help to set up the vagrant thing so that I can execute the code directly, not just via the unit test to make sure this refactored code above will work as expected, then I'll update the relevant unit tests.

-------

### What's next?

As this is a VIT issue, I think we can merge this PR now as the Codecov checks I mentioned above isn't a drama, and do the code refactoring as well as updating unit tests later.